engine/src/main/java/io/zeebe/engine/nwe/behavior/BpmnStateTransitionBehavior.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.engine.nwe.behavior;
+
+import io.zeebe.engine.metrics.WorkflowEngineMetrics;
+import io.zeebe.engine.nwe.BpmnElementContainerProcessor;
+import io.zeebe.engine.nwe.BpmnElementContext;
+import io.zeebe.engine.nwe.WorkflowInstanceStateTransitionGuard;
+import io.zeebe.engine.processor.KeyGenerator;
+import io.zeebe.engine.processor.TypedStreamWriter;
+import io.zeebe.engine.processor.workflow.WorkflowInstanceLifecycle;
+import io.zeebe.engine.processor.workflow.deployment.model.element.ExecutableFlowElement;
+import io.zeebe.engine.processor.workflow.deployment.model.element.ExecutableFlowNode;
+import io.zeebe.engine.processor.workflow.deployment.model.element.ExecutableSequenceFlow;
+import io.zeebe.engine.state.instance.ElementInstance;
+import io.zeebe.protocol.record.intent.WorkflowInstanceIntent;
+import io.zeebe.protocol.record.value.BpmnElementType;
+import java.util.function.Function;
+
+public final class BpmnStateTransitionBehavior {
+
+  private final TypedStreamWriter streamWriter;
+  private final KeyGenerator keyGenerator;
+  private final BpmnStateBehavior stateBehavior;
+  private final Function<BpmnElementType, BpmnElementContainerProcessor<ExecutableFlowElement>>
+      processorLookUp;
+
+  private final WorkflowInstanceStateTransitionGuard stateTransitionGuard;
+  private final WorkflowEngineMetrics metrics;
+
+  public BpmnStateTransitionBehavior(
+      final TypedStreamWriter streamWriter,
+      final KeyGenerator keyGenerator,
+      final BpmnStateBehavior stateBehavior,
+      final WorkflowEngineMetrics metrics,
+      final WorkflowInstanceStateTransitionGuard stateTransitionGuard,
+      final Function<BpmnElementType, BpmnElementContainerProcessor<ExecutableFlowElement>>
+          processorLookUp) {
+    this.streamWriter = streamWriter;
+    this.keyGenerator = keyGenerator;
+    this.stateBehavior = stateBehavior;
+    this.metrics = metrics;
+    this.stateTransitionGuard = stateTransitionGuard;
+    this.processorLookUp = processorLookUp;
+  }
+
+  public void transitionToActivated(final BpmnElementContext context) {
+    if (!WorkflowInstanceLifecycle.canTransition(
+        context.getIntent(), WorkflowInstanceIntent.ELEMENT_ACTIVATED)) {
+      throw new IllegalStateTransitionException(WorkflowInstanceIntent.ELEMENT_ACTIVATED, context);
+    }
+
+    transitionTo(context, WorkflowInstanceIntent.ELEMENT_ACTIVATED);
+
+    stateTransitionGuard.registerStateTransition(context, WorkflowInstanceIntent.ELEMENT_ACTIVATED);
```

Comment:
_Just asking:_

Wouldn't it be possible to define the state machine once in a lifecycle processor.
Currently you need to check whether the transition is OK and also you always need to take the transition in the elementTypeProcessor implementations. But this lifecycle is always the same right? 

Could look like this:
```
LifecycleProcessor {

 if (elementType == Activating)
{
  processors.get(elementType).onActivating(context)
  transitionToActivated()
}
else if (elementType == Completing)
{
  processors.get(elementType).onCompleting(context)
  transitionToCompleted()
}
}
```

Reply:
It would be nice but no :sweat_smile: See the comments above.